ykllvmwrap/src/jitbuilder.cc

```diff
@@ -0,0 +1,315 @@
+        // Skip first argument to start_tracing.
+        for (auto Arg = CI->arg_begin() + 1; Arg != CI->arg_end(); Arg++) {
+          Vec.push_back(Arg->get());
+        }
+        found = true;
+        break;
+      }
+    }
+  }
+  if (!found)
+    errx(EXIT_FAILURE, "failed to find trace inputs");
+  return Vec;
+}
+
+class JITBuilder {
+  // Global variables/functions that were copied over and need to be
+  // initialised.
+  vector<GlobalVariable *> cloned_globals;
+  // The new module that is being build.
+  Module *JITMod;
+  // A pointer to the call to YKTRACE_START in the AOT module (once
+  // encountered). When this changes from NULL to non-NULL, then we start
+  // copying instructions from the AOT module into the JIT module.
+  Instruction *StartTracingInstr = nullptr;
+
+public:
+  // The function name of this trace.
+  string TraceName;
+  // Mapping from AOT instructions to JIT instructions.
+  ValueToValueMapTy VMap;
+#ifndef NDEBUG
+  // Reverse mapping for debugging.
+  ValueToValueMapTy RevVMap;
+#endif
+
+  Module *createModule(char *FuncNames[], size_t BBs[], size_t Len,
+                       Module *AOTMod) {
+    LLVMContext &JITContext = AOTMod->getContext();
+    JITMod = new Module("", JITContext);
+    uint64_t TraceIdx = NextTraceIdx.fetch_add(1);
+    if (TraceIdx == numeric_limits<uint64_t>::max())
+      errx(EXIT_FAILURE, "trace index counter overflowed");
+
+    // Get var args from start_tracing call.
+    auto Inputs = getTraceInputs(AOTMod->getFunction(FuncNames[0]), BBs[0]);
+
+    std::vector<Type *> InputTypes;
+    for (auto Val : Inputs) {
+      InputTypes.push_back(Val->getType());
+    }
+
+    // Create function to store compiled trace.
+    TraceName = string(TRACE_FUNC_PREFIX) + to_string(TraceIdx);
+    llvm::FunctionType *FType =
+        llvm::FunctionType::get(Type::getVoidTy(JITContext), InputTypes, false);
+    llvm::Function *JITFunc = llvm::Function::Create(
+        FType, Function::InternalLinkage, TraceName, JITMod);
+    JITFunc->setCallingConv(CallingConv::C);
+
+    // Create entry block and setup builder.
+    auto DstBB = BasicBlock::Create(JITContext, "", JITFunc);
+    llvm::IRBuilder<> Builder(JITContext);
+    Builder.SetInsertPoint(DstBB);
+
+    // Variables that are used (but not defined) inbetween start and stop
+    // tracing need to be replaced with function arguments which the user passes
+    // into the compiled trace. This loop creates a mapping from those original
+    // variables to the function arguments of the compiled trace function.
+    for (size_t Idx = 0; Idx != Inputs.size(); Idx++) {
+      Value *OldVal = Inputs[Idx];
+      Value *NewVal = JITFunc->getArg(Idx);
+      assert(NewVal->getType()->isPointerTy());
+      VMap[OldVal] = NewVal;
+    }
+
+    std::vector<CallInst *> inlined_calls;
+    CallInst *last_call = nullptr;
+
+    // Iterate over the PT trace and stitch together all traced blocks.
```

Comment:
We shouldn't mention PT here, as the software tracer would share the same code, right?

Reply:
Fixed in 9ff6b8f.